### PR TITLE
zinc: init at version 0.3.5

### PR DESCRIPTION
**Pull request by mfenniak**
###### Description of changes

Adds package and NixOS module for ZincSearch (https://github.com/zinclabs/zinc), a lightweight alternative to ElasticSearch that requires minimal server resources.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
    - [x] new NixOS tests added
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
  - [x] Tested by hand in x86_64-linux, and, via new NixOS tests
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/search/zinc/default.nix [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    export ZINC_FIRST_ADMIN_PASSWORD=Complexpass#123
+  '';
+
+  installPhase = ''
+    mkdir -p $out/bin
+    cp zinc $out/bin/
+  '';
+
+  meta = with lib; {
+    inherit (src.meta) homepage;
```
Comment (SuperSandro2000) [inline pkgs/servers/search/zinc/default.nix:64]: Please write the homepage in plain text

File: pkgs/servers/search/zinc/default.nix [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    description = "ZincSearch - A lightweight alternative to elasticsearch that requires minimal resources";
```
Comment (SuperSandro2000) [inline pkgs/servers/search/zinc/default.nix:65]:
```suggestion
    description = "Lightweight alternative to elasticsearch that requires minimal resources";
```

File: pkgs/servers/search/zinc/default.nix [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  vendorSha256 = "sha256-akjb0cxHbITKS26c+7lVSHWO/KRoQVVKzAOra+tdAD8=";
+  preBuild = ''
+    # (can't symlink; "can't embed irregular file")
+    cp -rf ${web} web/dist
+  '';
+  buildPhase = ''
+    runHook preBuild
+    go build \
+      -o zinc \
+      -ldflags="-s -w -X github.com/zinclabs/zinc/pkg/meta.Version=${version} -X github.com/zinclabs/zinc/pkg/meta.CommitHash=${src.rev} -X github.com/zinclabs/zinc/pkg/meta.BuildDate=19691231" \
```
Comment (SuperSandro2000) [inline pkgs/servers/search/zinc/default.nix:48]: Please use nixos default date for the build date

File: pkgs/servers/search/zinc/default.nix [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      npm run build
+    '';
+    installPhase = ''
+      cp -r dist $out/
+    '';
+  };
+
+  vendorSha256 = "sha256-akjb0cxHbITKS26c+7lVSHWO/KRoQVVKzAOra+tdAD8=";
+  preBuild = ''
+    # (can't symlink; "can't embed irregular file")
```
Comment (SuperSandro2000) [inline pkgs/servers/search/zinc/default.nix:41]:
```suggestion
    # goembed does not support symlinks
```

File: pkgs/servers/search/zinc/default.nix [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+  # refresh web packaging by:
+  #   git clone zinc
+  #   cd zinc/web
+  #   node2nix -l ./package-lock.json -d
+  #   cp *.nix ~/Dev/custom-nixpkgs/zinc/web/
+  web = stdenv.mkDerivation {
+    pname = "${pname}-web";
+    inherit src version;
+    buildInputs = [pkgs.nodejs-18_x];
```
Comment (SuperSandro2000) [inline pkgs/servers/search/zinc/default.nix:27]:
```suggestion
    buildInputs = [ nodejs-18_x ];
```

File: pkgs/servers/search/zinc/default.nix [hunk 6]
```diff
@@ -0,0 +1,70 @@
+{ lib, stdenv, fetchFromGitHub, pkgs, buildGoModule }:
```
Comment (SuperSandro2000) [inline pkgs/servers/search/zinc/default.nix:1]:
```suggestion
{ lib, stdenv, fetchFromGitHub, buildGoModule }:
```
Please do not use pkgs here

File: pkgs/servers/search/zinc/default.nix [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+let
+  nodeDependencies = (pkgs.callPackage ./web/default.nix {}).nodeDependencies.override (old: {
```
Comment (SuperSandro2000) [inline pkgs/servers/search/zinc/default.nix:4]:
```suggestion
  nodeDependencies = (pkgs.callPackage ./web {}).nodeDependencies.override (old: {
```

File: nixos/modules/services/search/zinc.nix [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        ZINC_SERVER_PORT = toString cfg.port;
+        ZINC_PROMETHEUS_ENABLE = "true";
+        GIN_MODE = "release"; # https://github.com/zinclabs/zinc/blob/16ccff60aa361421e0075e44e47142986cedebe1/pkg/config/config.go#L121
+      };
+
+      script = ''
+        export ZINC_DATA_PATH=$STATE_DIRECTORY
+        export ZINC_FIRST_ADMIN_PASSWORD="$(head -n1 ${escapeShellArg cfg.firstAdminPasswordFile})"
+        if [[ $ZINC_FIRST_ADMIN_PASSWORD == "" ]]; then
+          echo "ZINC_FIRST_ADMIN_PASSWORD is unset"
```
Comment (SuperSandro2000) [inline nixos/modules/services/search/zinc.nix:67]: This should mention what to do.

File: nixos/modules/services/search/zinc.nix [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+  config = mkIf cfg.enable {
+    systemd.services.zinc = {
+      description = "ZincSearch Daemon";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      environment = {
+        ZINC_FIRST_ADMIN_USER = cfg.firstAdminUsername;
+        ZINC_SERVER_PORT = toString cfg.port;
+        ZINC_PROMETHEUS_ENABLE = "true";
+        GIN_MODE = "release"; # https://github.com/zinclabs/zinc/blob/16ccff60aa361421e0075e44e47142986cedebe1/pkg/config/config.go#L121
```
Comment (SuperSandro2000) [inline nixos/modules/services/search/zinc.nix:60]: The comment fails to explain why though

File: nixos/modules/services/search/zinc.nix [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+    openFirewall = mkOption {
+      type = types.bool;
+      default = false;
+      description = lib.mdDoc "Open ports in the firewall for ZincSearch.";
+    };
+  };
+
+  ###### implementation
+
```
Comment (SuperSandro2000) [inline nixos/modules/services/search/zinc.nix:50]:
```suggestion
```

File: nixos/modules/services/search/zinc.nix [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+in
+{
+  ###### interface
+
+  options.services.zinc = {
+    enable = mkOption {
+      type = types.bool;
+      description = lib.mdDoc "Whether to enable ZincSearch.";
+      default = false;
+    };
```
Comment (SuperSandro2000) [inline nixos/modules/services/search/zinc.nix:16]: Please use mkEnableOption

File: nixos/modules/services/search/zinc.nix [hunk 12]
```diff
@@ -0,0 +1,110 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.zinc;
+in
+{
+  ###### interface
+
```
Comment (SuperSandro2000) [inline nixos/modules/services/search/zinc.nix:10]:
```suggestion
```

**Comment by mfenniak** [comment]
Thanks for the review @SuperSandro2000; all comments are addressed.

**Comment by mfenniak** [comment]
Closing as this package has been added in #197566.  This PR does have a nixos module which could be resurrected if someone wished.

**Pull request closed without merging**